File: 2015/06_Probably_a_Fire_Hazard/Day06.kt
// Day 6: Probably a Fire Hazard
// https://adventofcode.com/2015/day/6

import java.nio.file.Files
import java.nio.file.Paths

fun readInput(name: String): String {
    return Files.readString(Paths.get("$name.txt")).trim()
}

data class Instruction(val action: String, val x1: Int, val y1: Int, val x2: Int, val y2: Int)

fun prepareData(input: String): List<Instruction> {
    val pattern = Regex("(turn on|turn off|toggle) (\\d+),(\\d+) through (\\d+),(\\d+)")
    return input.split("\r\n").map { line ->
        val matchResult = pattern.find(line)!!
        val (action, x1, y1, x2, y2) = matchResult.destructured
        Instruction(action, x1.toInt(), y1.toInt(), x2.toInt(), y2.toInt())
    }
}

fun part1(input: String): Int {
    val instructions = prepareData(input)
    val grid = Array(1000) { IntArray(1000) }

    for ((action, x1, y1, x2, y2) in instructions) {
        for (x in x1..x2) {
            for (y in y1..y2) {
                when (action) {
                    "turn on" -> grid[x][y] = 1
                    "turn off" -> grid[x][y] = 0
                    "toggle" -> grid[x][y] = if (grid[x][y] == 1) 0 else 1
                }
            }
        }
    }

    return grid.sumOf { row -> row.sum() }
}

fun part2(input: String): Int {
    val instructions = prepareData(input)
    val grid = Array(1000) { IntArray(1000) }

    for ((action, x1, y1, x2, y2) in instructions) {
        for (x in x1..x2) {
            for (y in y1..y2) {
                when (action) {
                    "turn on" -> grid[x][y]++
                    "turn off" -> if (grid[x][y] > 0) grid[x][y]--
                    "toggle" -> grid[x][y] += 2
                }
            }
        }
    }

    return grid.sumOf { row -> row.sum() }
}

fun main() {
    val day = "Day06"
    val input = readInput(day)
    val inputTest = readInput("${day}_test")

    println("Part 1: ${part1(input)}")
    println("Part 1: ${part1(inputTest)}")

    println("Part 2: ${part2(input)}")
    println("Part 2: ${part2(inputTest)}")
}

File: 2015/06_Probably_a_Fire_Hazard/Day06.js
// Day 6: Probably a Fire Hazard
// https://adventofcode.com/2015/day/6

const utils = require('../../utils');

function prepareData(input) {
  const pattern = /(turn on|turn off|toggle) (\d+),(\d+) through (\d+),(\d+)/;
  return input.split('\r\n').map((line) => {
    const [, action, x1, y1, x2, y2] = line.match(pattern);
    return {
      action,
      x1: Number(x1),
      y1: Number(y1),
      x2: Number(x2),
      y2: Number(y2),
    };
  });
}

function part1(input) {
  const instructions = prepareData(input);
  const grid = Array.from({ length: 1000 }, () => new Array(1000).fill(0));
  for (const { action, x1, x2, y1, y2 } of instructions) {
    for (let x = x1; x <= x2; x++) {
      for (let y = y1; y <= y2; y++) {
        if (action === 'turn on') {
          grid[x][y] = 1;
        } else if (action === 'turn off') {
          grid[x][y] = 0;
        } else if (action === 'toggle') {
          grid[x][y] ^= 1;
        }
      }
    }
  }

  return grid.flat().reduce((acc, val) => acc + val, 0);
}

function part2(input) {
  const instructions = prepareData(input);
  const grid = Array.from({ length: 1000 }, () => new Array(1000).fill(0));
  for (const { action, x1, x2, y1, y2 } of instructions) {
    for (let x = x1; x <= x2; x++) {
      for (let y = y1; y <= y2; y++) {
        if (action === 'turn on') {
          grid[x][y]++;
        } else if (action === 'turn off') {
          if (grid[x][y] > 0) grid[x][y]--;
        } else if (action === 'toggle') {
          grid[x][y] += 2;
        }
      }
    }
  }

  return grid.flat().reduce((acc, val) => acc + val, 0);
}

const day = 6;
const input = utils.readInputByDay(day);
const inputTest = utils.readInputByDay(day, '_test');

console.log('Part 1:', part1(input)); // 998996
console.time('Part 1');
console.log('Part 1:', part1(inputTest)); // 569999
console.timeEnd('Part 1');

console.log('Part 2:', part2(input)); // 1001996
console.time('Part 2');
console.log('Part 2:', part2(inputTest)); // 17836115
console.timeEnd('Part 2');
